lib/widgets/compose_box.dart

```diff
@@ -44,21 +47,75 @@ class TopicTextEditingController extends TextEditingController {
 
 enum ContentValidationError {
   empty,
-  tooLong;
+  tooLong,
+  uploadInProgress;
 
-  // Later: upload in progress; quote-and-reply in progress
+  // Later: quote-and-reply in progress
 
   String message() {
     switch (this) {
       case ContentValidationError.tooLong:
         return "Message length shouldn't be greater than 10000 characters.";
       case ContentValidationError.empty:
         return 'You have nothing to send!';
+      case ContentValidationError.uploadInProgress:
+        return 'Please wait for the upload to complete.';
     }
   }
 }
 
 class ContentTextEditingController extends TextEditingController {
+  int _uploadCounter = 0;
+
+  final Map<int, ({String filename, String placeholder})> _uploads = {};
+
+  /// The cursor position, if selection is valid and collapsed, else text.length.
+  TextRange _cursorPositionOrEnd() {
+    final TextRange selection = value.selection;
+    final String text = value.text;
+    final int index = (selection.isValid && selection.end < text.length && selection.isCollapsed)
```

Comment:
Probably it doesn't change behavior in this case? I think I saw it as critical for deciding if "[the] selection is valid" but noticed it's missing in [`selection.isValid`](https://api.flutter.dev/flutter/dart-ui/TextRange/isValid.html).

Reply:
I see. Yeah, there should be an invariant that `selection.end <= text.length` (but I think *equal* to the length is fine and normal — that's what you have when the caret is at the end.) And then there's no need for us to re-check invariants that the framework should be providing.

The `TextSelection` API is unusual in that it has this special value it calls "invalid", namely (-1,-1), and it actually uses that "invalid" value. That should really be represented as null instead, so that an actual `TextSelection` value would always mean a valid selection. Everyone involved understands that the current API is unfortunate, but it's been a pain to try to migrate off of:
https://github.com/flutter/flutter/pull/122480#discussion_r1142323494
https://github.com/flutter/flutter/issues/95978#issuecomment-1006997846
https://github.com/flutter/flutter/pull/79615

So when dealing with that `isValid` getter, one should think of it as having a meaning similar to "!= null". It fortunately doesn't open up a need to start checking other "validity" conditions of that API where one otherwise wouldn't.